facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+                    "Bearer {}",
+                    self.ensure_default_service_account_oauth_token()?
+                ),
+            )
+            .set("Content-Type", "application/json")
+            // At the moment, these tokens are only used to read and write from
+            // GCS buckets, so we request an appropriate scope. We could further
+            // narrow this to a devstorage.read token if we knew what it was
+            // going to be used for, but that would require this object to
+            // manage two impersonation tokens which is more work than I want to
+            // do right now.
+            // https://cloud.google.com/storage/docs/authentication#oauth-scopes
+            .send_json(ureq::json!({
+                "scope": [
+                    "https://www.googleapis.com/auth/devstorage.read_write"
+                ]
+            }));
+        if http_response.error() {
+            return Err(anyhow!(
+                "failed to get Oauth token to impersonate service account {}: {:?}",
+                service_account_to_impersonate,
+                http_response
+            ));
+        }
+
+        let response = http_response
+            .into_json_deserialize::<GenerateAccessTokenResponse>()
+            .context("failed to deserialize response from IAM API")?;
+        self.impersonated_service_account_oauth_token = Some(OauthToken {
+            token: response.access_token.clone(),
+            expiration: response.expire_time,
+        });
+
+        Ok(response.access_token)
+    }
+}
+
+// GCSTransport manages reading and writing from GCS buckets. Currently, the
+// only authentication it supports is by means of impersonating a specified GCP
+// service account, which is assumed to have the necessary read or write
+// permissions on the bucket.
+pub struct GCSTransport {
+    path: GCSPath,
+    oauth_token_provider: OauthTokenProvider,
+}
+
+impl GCSTransport {
+    pub fn new(path: GCSPath, impersonate: Option<String>) -> GCSTransport {
+        GCSTransport {
+            path: path.ensure_directory_prefix(),
+            oauth_token_provider: OauthTokenProvider::new(impersonate),
+        }
+    }
+}
+
+impl Transport for GCSTransport {
+    fn get(&mut self, key: &str) -> Result<Box<dyn Read>> {
+        // Per API reference, the object key must be URL encoded.
+        // API reference: https://cloud.google.com/storage/docs/json_api/v1/objects/get
+        let encoded_key = urlencoding::encode(&[&self.path.key, key].concat());
+        let url = format!(
+            "{}/storage/v1/b/{}/o/{}",
+            STORAGE_API_BASE_URL, self.path.bucket, encoded_key
+        );
+
+        let response = ureq::get(&url)
+            // Ensures response body will be content and not JSON metadata.
+            // https://cloud.google.com/storage/docs/json_api/v1/objects/get#parameters
+            .query("alt", "media")
+            .set(
+                "Authorization",
+                &format!(
```

Comment:
Style: when I find myself with cascading fields like this I try to break them out into variables, e.g.

```
let auth = format!(...);
...
.set("Authorization", auth);
```